// ng-click-outside/lib_commonjs/click-outside.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClickOutsideModule = void 0;
var core_1 = require("@angular/core");
var click_outside_directive_1 = require("./click-outside.directive");
var i0 = require("@angular/core");
var ClickOutsideModule = (function () {
    function ClickOutsideModule() {
    }
    ClickOutsideModule.ɵmod = i0.ɵɵdefineNgModule({ type: ClickOutsideModule });
    ClickOutsideModule.ɵinj = i0.ɵɵdefineInjector({ factory: function ClickOutsideModule_Factory(t) { return new (t || ClickOutsideModule)(); } });
    return ClickOutsideModule;
}());
exports.ClickOutsideModule = ClickOutsideModule;
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(ClickOutsideModule, { declarations: [click_outside_directive_1.ClickOutsideDirective], exports: [click_outside_directive_1.ClickOutsideDirective] }); })();
(function () { i0.ɵsetClassMetadata(ClickOutsideModule, [{
        type: core_1.NgModule,
        args: [{
                declarations: [click_outside_directive_1.ClickOutsideDirective],
                exports: [click_outside_directive_1.ClickOutsideDirective]
            }]
    }], null, null); })();
