// ngx-gauge/bundles/ngx-gauge.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ngx-gauge', ['exports', '@angular/core', '@angular/common'], factory) :
    (global = global || self, factory(global['ngx-gauge'] = {}, global.ng.core, global.ng.common));
}(this, (function (exports, core, common) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m) return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function clamp(value, min, max) {
        return Math.max(min, Math.min(max, value));
    }
    function coerceBooleanProperty(value) {
        return value != null && "" + value !== 'false';
    }
    function coerceNumberProperty(value, fallbackValue) {
        if (fallbackValue === void 0) { fallbackValue = 0; }
        return isNaN(parseFloat(value)) || isNaN(Number(value)) ? fallbackValue : Number(value);
    }
    function cssUnit(value) {
        return value + "px";
    }
    function isNumber(value) {
        return value != undefined && !isNaN(parseFloat(value)) && !isNaN(Number(value));
    }

    var NgxGaugeAppend = /** @class */ (function () {
        function NgxGaugeAppend() {
        }
        NgxGaugeAppend = __decorate([
            core.Directive({
                selector: "ngx-gauge-append",
                exportAs: "ngxGaugeAppend"
            })
        ], NgxGaugeAppend);
        return NgxGaugeAppend;
    }());
    var NgxGaugePrepend = /** @class */ (function () {
        function NgxGaugePrepend() {
        }
        NgxGaugePrepend = __decorate([
            core.Directive({
                selector: "ngx-gauge-prepend",
                exportAs: "ngxGaugePrepend"
            })
        ], NgxGaugePrepend);
        return NgxGaugePrepend;
    }());
    var NgxGaugeValue = /** @class */ (function () {
        function NgxGaugeValue() {
        }
        NgxGaugeValue = __decorate([
            core.Directive({
                selector: "ngx-gauge-value",
                exportAs: "ngxGaugeValue"
            })
        ], NgxGaugeValue);
        return NgxGaugeValue;
    }());
    var NgxGaugeLabel = /** @class */ (function () {
        function NgxGaugeLabel() {
        }
        NgxGaugeLabel = __decorate([
            core.Directive({
                selector: "ngx-gauge-label",
                exportAs: "ngxGaugeLabel"
            })
        ], NgxGaugeLabel);
        return NgxGaugeLabel;
    }());

    var DEFAULTS = {
        MIN: 0,
        MAX: 100,
        TYPE: 'arch',
        THICK: 4,
        FOREGROUND_COLOR: 'rgba(0, 150, 136, 1)',
        BACKGROUND_COLOR: 'rgba(0, 0, 0, 0.1)',
        CAP: 'butt',
        SIZE: 200
    };
    var NgxGauge = /** @class */ (function () {
        function NgxGauge(_elementRef, _renderer) {
            this._elementRef = _elementRef;
            this._renderer = _renderer;
            this._size = DEFAULTS.SIZE;
            this._min = DEFAULTS.MIN;
            this._max = DEFAULTS.MAX;
            this._animate = true;
            this._initialized = false;
            this._animationRequestID = 0;
            this.ariaLabel = '';
            this.ariaLabelledby = null;
            this.type = DEFAULTS.TYPE;
            this.cap = DEFAULTS.CAP;
            this.thick = DEFAULTS.THICK;
            this.foregroundColor = DEFAULTS.FOREGROUND_COLOR;
            this.backgroundColor = DEFAULTS.BACKGROUND_COLOR;
            this.thresholds = Object.create(null);
            this._value = 0;
            this.duration = 1200;
        }
        Object.defineProperty(NgxGauge.prototype, "size", {
            get: function () { return this._size; },
            set: function (value) {
                this._size = coerceNumberProperty(value);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NgxGauge.prototype, "min", {
            get: function () { return this._min; },
            set: function (value) {
                this._min = coerceNumberProperty(value, DEFAULTS.MIN);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NgxGauge.prototype, "animate", {
            get: function () { return this._animate; },
            set: function (value) {
                this._animate = coerceBooleanProperty(value);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NgxGauge.prototype, "max", {
            get: function () { return this._max; },
            set: function (value) {
                this._max = coerceNumberProperty(value, DEFAULTS.MAX);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NgxGauge.prototype, "value", {
            get: function () { return this._value; },
            set: function (val) {
                this._value = coerceNumberProperty(val);
            },
            enumerable: true,
            configurable: true
        });
        NgxGauge.prototype.ngOnChanges = function (changes) {
            var isCanvasPropertyChanged = changes['thick'] || changes['type'] || changes['cap'] || changes['size'];
            var isDataChanged = changes['value'] || changes['min'] || changes['max'];
            if (this._initialized) {
                if (isDataChanged) {
                    var nv = void 0, ov = void 0;
                    if (changes['value']) {
                        nv = changes['value'].currentValue;
                        ov = changes['value'].previousValue;
                    }
                    this._update(nv, ov);
                }
                if (isCanvasPropertyChanged) {
                    this._destroy();
                    this._init();
                }
            }
        };
        NgxGauge.prototype._updateSize = function () {
            this._renderer.setStyle(this._elementRef.nativeElement, 'width', cssUnit(this._getWidth()));
            this._renderer.setStyle(this._elementRef.nativeElement, 'height', cssUnit(this._getCanvasHeight()));
            this._canvas.nativeElement.width = this._getWidth();
            this._canvas.nativeElement.height = this._getCanvasHeight();
            this._renderer.setStyle(this._label.nativeElement, 'transform', 'translateY(' + (this.size / 3 * 2 - this.size / 13 / 4) + 'px)');
            this._renderer.setStyle(this._reading.nativeElement, 'transform', 'translateY(' + (this.size / 2 - this.size * 0.22 / 2) + 'px)');
        };
        NgxGauge.prototype.ngAfterViewInit = function () {
            if (this._canvas) {
                this._init();
            }
        };
        NgxGauge.prototype.ngOnDestroy = function () {
            this._destroy();
        };
        NgxGauge.prototype._getBounds = function (type) {
            var head, tail;
            if (type == 'semi') {
                head = Math.PI;
                tail = 2 * Math.PI;
            }
            else if (type == 'full') {
                head = 1.5 * Math.PI;
                tail = 3.5 * Math.PI;
            }
            else if (type === 'arch') {
                head = 0.8 * Math.PI;
                tail = 2.2 * Math.PI;
            }
            return { head: head, tail: tail };
        };
        NgxGauge.prototype._drawShell = function (start, middle, tail, color) {
            var center = this._getCenter(), radius = this._getRadius();
            middle = Math.max(middle, start); // never below 0%
            middle = Math.min(middle, tail); // never exceed 100%
            if (this._initialized) {
                this._clear();
                this._context.beginPath();
                this._context.strokeStyle = this.backgroundColor;
                this._context.arc(center.x, center.y, radius, middle, tail, false);
                this._context.stroke();
                this._context.beginPath();
                this._context.strokeStyle = color;
                this._context.arc(center.x, center.y, radius, start, middle, false);
                this._context.stroke();
            }
        };
        NgxGauge.prototype._clear = function () {
            this._context.clearRect(0, 0, this._getWidth(), this._getHeight());
        };
        NgxGauge.prototype._getWidth = function () {
            return this.size;
        };
        NgxGauge.prototype._getHeight = function () {
            return this.size;
        };
        // canvas height will be shorter for type 'semi' and 'arch'
        NgxGauge.prototype._getCanvasHeight = function () {
            return (this.type == 'arch' || this.type == 'semi')
                ? 0.85 * this._getHeight()
                : this._getHeight();
        };
        NgxGauge.prototype._getRadius = function () {
            var center = this._getCenter();
            return center.x - this.thick;
        };
        NgxGauge.prototype._getCenter = function () {
            var x = this._getWidth() / 2, y = this._getHeight() / 2;
            return { x: x, y: y };
        };
        NgxGauge.prototype._init = function () {
            this._context = this._canvas.nativeElement.getContext('2d');
            this._initialized = true;
            this._updateSize();
            this._setupStyles();
            this._create();
        };
        NgxGauge.prototype._destroy = function () {
            if (this._animationRequestID) {
                window.cancelAnimationFrame(this._animationRequestID);
                this._animationRequestID = 0;
            }
            this._clear();
            this._context = null;
            this._initialized = false;
        };
        NgxGauge.prototype._setupStyles = function () {
            this._context.lineCap = this.cap;
            this._context.lineWidth = this.thick;
        };
        NgxGauge.prototype._getForegroundColorByRange = function (value) {
            var match = Object.keys(this.thresholds)
                .filter(function (item) { return isNumber(item) && Number(item) <= value; })
                .sort(function (a, b) { return Number(a) - Number(b); })
                .reverse()[0];
            return match !== undefined
                ? this.thresholds[match].color || this.foregroundColor
                : this.foregroundColor;
        };
        NgxGauge.prototype._create = function (nv, ov) {
            var self = this, type = this.type, bounds = this._getBounds(type), duration = this.duration, min = this.min, max = this.max, value = clamp(this.value, this.min, this.max), start = bounds.head, unit = (bounds.tail - bounds.head) / (max - min), displacement = unit * (value - min), tail = bounds.tail, color = this._getForegroundColorByRange(value), startTime;
            if (self._animationRequestID) {
                window.cancelAnimationFrame(self._animationRequestID);
            }
            function animate(timestamp) {
                timestamp = timestamp || new Date().getTime();
                var runtime = timestamp - startTime;
                var progress = Math.min(runtime / duration, 1);
                var previousProgress = ov ? (ov - min) * unit : 0;
                var middle = start + previousProgress + displacement * progress;
                self._drawShell(start, middle, tail, color);
                if (self._animationRequestID && (runtime < duration)) {
                    self._animationRequestID = window.requestAnimationFrame(function (timestamp) { return animate(timestamp); });
                }
                else {
                    window.cancelAnimationFrame(self._animationRequestID);
                }
            }
            if (this._animate) {
                if (nv != undefined && ov != undefined) {
                    displacement = unit * nv - unit * ov;
                }
                self._animationRequestID = window.requestAnimationFrame(function (timestamp) {
                    startTime = timestamp || new Date().getTime();
                    animate(startTime);
                });
            }
            else {
                self._drawShell(start, start + displacement, tail, color);
            }
        };
        NgxGauge.prototype._update = function (nv, ov) {
            this._clear();
            this._create(nv, ov);
        };
        NgxGauge.ctorParameters = function () { return [
            { type: core.ElementRef },
            { type: core.Renderer2 }
        ]; };
        __decorate([
            core.ViewChild('canvas', { static: true }),
            __metadata("design:type", core.ElementRef)
        ], NgxGauge.prototype, "_canvas", void 0);
        __decorate([
            core.ViewChild('rLabel', { static: true }),
            __metadata("design:type", core.ElementRef)
        ], NgxGauge.prototype, "_label", void 0);
        __decorate([
            core.ViewChild('reading', { static: true }),
            __metadata("design:type", core.ElementRef)
        ], NgxGauge.prototype, "_reading", void 0);
        __decorate([
            core.ContentChild(NgxGaugeLabel),
            __metadata("design:type", NgxGaugeLabel)
        ], NgxGauge.prototype, "_labelChild", void 0);
        __decorate([
            core.ContentChild(NgxGaugePrepend),
            __metadata("design:type", NgxGaugePrepend)
        ], NgxGauge.prototype, "_prependChild", void 0);
        __decorate([
            core.ContentChild(NgxGaugeAppend),
            __metadata("design:type", NgxGaugeAppend)
        ], NgxGauge.prototype, "_appendChild", void 0);
        __decorate([
            core.ContentChild(NgxGaugeValue),
            __metadata("design:type", NgxGaugeValue)
        ], NgxGauge.prototype, "_valueDisplayChild", void 0);
        __decorate([
            core.Input('aria-label'),
            __metadata("design:type", String)
        ], NgxGauge.prototype, "ariaLabel", void 0);
        __decorate([
            core.Input('aria-labelledby'),
            __metadata("design:type", String)
        ], NgxGauge.prototype, "ariaLabelledby", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Number),
            __metadata("design:paramtypes", [Number])
        ], NgxGauge.prototype, "size", null);
        __decorate([
            core.Input(),
            __metadata("design:type", Number),
            __metadata("design:paramtypes", [Number])
        ], NgxGauge.prototype, "min", null);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean),
            __metadata("design:paramtypes", [Object])
        ], NgxGauge.prototype, "animate", null);
        __decorate([
            core.Input(),
            __metadata("design:type", Number),
            __metadata("design:paramtypes", [Number])
        ], NgxGauge.prototype, "max", null);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxGauge.prototype, "type", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxGauge.prototype, "cap", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Number)
        ], NgxGauge.prototype, "thick", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxGauge.prototype, "label", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxGauge.prototype, "append", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxGauge.prototype, "prepend", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxGauge.prototype, "foregroundColor", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxGauge.prototype, "backgroundColor", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Object)
        ], NgxGauge.prototype, "thresholds", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Number),
            __metadata("design:paramtypes", [Number])
        ], NgxGauge.prototype, "value", null);
        __decorate([
            core.Input(),
            __metadata("design:type", Number)
        ], NgxGauge.prototype, "duration", void 0);
        NgxGauge = __decorate([
            core.Component({
                selector: 'ngx-gauge',
                template: "<div class=\"reading-block\" #reading [style.fontSize]=\"size * 0.22 + 'px'\">\r\n  <!-- This block can not be indented correctly, because line breaks cause layout spacing, related problem: https://pt.stackoverflow.com/q/276760/2998 -->\r\n  <u class=\"reading-affix\" [ngSwitch]=\"_prependChild != null\"><ng-content select=\"ngx-gauge-prepend\" *ngSwitchCase=\"true\"></ng-content><ng-container *ngSwitchCase=\"false\">{{prepend}}</ng-container></u><ng-container [ngSwitch]=\"_valueDisplayChild != null\"><ng-content *ngSwitchCase=\"true\" select=\"ngx-gauge-value\"></ng-content><ng-container *ngSwitchCase=\"false\">{{value | number}}</ng-container></ng-container><u class=\"reading-affix\" [ngSwitch]=\"_appendChild != null\"><ng-content select=\"ngx-gauge-append\" *ngSwitchCase=\"true\"></ng-content><ng-container *ngSwitchCase=\"false\">{{append}}</ng-container></u>\r\n</div>\r\n<div class=\"reading-label\" #rLabel\r\n     [style.fontSize]=\"size / 13 + 'px'\"\r\n     [ngSwitch]=\"_labelChild != null\">\r\n  <ng-content select=\"ngx-gauge-label\" *ngSwitchCase=\"true\"></ng-content>\r\n  <ng-container *ngSwitchCase=\"false\">{{label}}</ng-container>\r\n</div>\r\n<canvas #canvas></canvas>\r\n",
                host: {
                    'role': 'slider',
                    'aria-readonly': 'true',
                    '[class.ngx-gauge-meter]': 'true',
                    '[attr.aria-valuemin]': 'min',
                    '[attr.aria-valuemax]': 'max',
                    '[attr.aria-valuenow]': 'value',
                    '[attr.aria-label]': 'ariaLabel',
                    '[attr.aria-labelledby]': 'ariaLabelledby'
                },
                encapsulation: core.ViewEncapsulation.None,
                styles: [".ngx-gauge-meter{display:inline-block;text-align:center;position:relative}.reading-block{position:absolute;width:100%;font-weight:400;white-space:nowrap;text-align:center;overflow:hidden;text-overflow:ellipsis}.reading-label{font-family:inherit;width:100%;display:inline-block;position:absolute;text-align:center;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;font-weight:400}.reading-affix{text-decoration:none;font-size:.6em;opacity:.8;font-weight:200;padding:0 .18em}.reading-affix:first-child{padding-left:0}.reading-affix:last-child{padding-right:0}"]
            }),
            __metadata("design:paramtypes", [core.ElementRef, core.Renderer2])
        ], NgxGauge);
        return NgxGauge;
    }());

    var NgxGaugeModule = /** @class */ (function () {
        function NgxGaugeModule() {
        }
        NgxGaugeModule = __decorate([
            core.NgModule({
                imports: [common.CommonModule],
                declarations: [NgxGauge, NgxGaugeAppend, NgxGaugePrepend, NgxGaugeValue, NgxGaugeLabel],
                exports: [NgxGauge, NgxGaugeAppend, NgxGaugePrepend, NgxGaugeValue, NgxGaugeLabel]
            })
        ], NgxGaugeModule);
        return NgxGaugeModule;
    }());

    exports.NgxGaugeModule = NgxGaugeModule;
    exports.ɵa = NgxGauge;
    exports.ɵb = NgxGaugeAppend;
    exports.ɵc = NgxGaugePrepend;
    exports.ɵd = NgxGaugeValue;
    exports.ɵe = NgxGaugeLabel;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
